File: Data Structures/LinkedLists/Single_LinkedList/Insertion/insertion.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def push_back(self, val):
        newNode = Node(val)
        if (self.head == None):
            self.head = newNode
            return
        temp = self.head
        while (temp.next != None):
            temp = temp.next
        temp.next = newNode

    def push_front(self, val):
        newNode = Node(val)
        if (self.head is None):
            self.head = newNode
            return
        newNode.next = self.head
        self.head = newNode

    def insert_after(self,key,val):
        newNode = Node(val)
        if(self.head.data == key):
            newNode.next = self.head.next
            self.head.next = newNode
            return
        temp = self.head
        while(temp.data != key):
            temp = temp.next
            if temp is None:
                return
        newNode.next = temp.next
        temp.next = newNode
    
    def print_list(self):
        temp = self.head
        while(temp is not None):
            print(temp.data,end=" ")
            temp = temp.next
        print("")


if __name__ == '__main__':
    list = LinkedList()
    ch = None
    while(ch!=0):
        print("1.pushback")
        print("2.pushfront")
        print("3.insertafter")
        print("4.print_list")
        ch=int(input())
        if ch == 1:
            print("enter value to add: ")
            val = int(input())
            list.push_back(val)
        elif ch == 2:
            print("enter value to add: ")
            val = int(input())
            list.push_front(val)
        elif ch == 3:
            print("enter value to add: ")
            val = int(input())
            print("enter key: ")
            key = int(input())
            list.insert_after(key,val)
        elif ch == 4:
            list.print_list()

File: Data Structures/LinkedLists/Single_LinkedList/Insertion/insertion.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node(int val)
    {
        data = val;
        next = NULL;
    }
};

void push_back(Node *&head, int val)
{
    Node *NewNode = new Node(val);
    if (head == NULL)
    {
        head = NewNode;
        return;
    }
    Node *temp = head;
    while (temp->next != NULL)
    {
        temp = temp->next;
    }
    temp->next = NewNode;
}

void push_front(Node *&head, int val)
{
    Node *NewNode = new Node(val);
    if (head == NULL)
    {
        head = NewNode;
        return;
    }
    NewNode->next = head;
    head = NewNode;
}

void push_after(Node *&head, int key, int val)
{
    Node *NewNode = new Node(val);
    if (head->data == key)
    {
        NewNode->next = head->next;
        head->next = NewNode;
        return;
    }
    Node *temp = head;
    while (temp->data != key)
    {
        temp = temp->next;
        if (temp == NULL)
        {
            return;
        }
    }
    NewNode->next = temp->next;
    temp->next = NewNode;
}

void display_list(Node *&head)
{
    if (head == NULL)
    {
        cout << "list is empty" << endl;
        return;
    }
    Node *temp = head;
    while (temp != NULL)
    {
        cout << temp->data << " -> ";
        temp = temp->next;
    }
    cout << endl;
}

int main()
{
    Node *head = NULL;
    int ch;
    while (ch != 0)
    {
        cout << "1. push_back(head,val)" << endl;
        cout << "2. display_list(head)" << endl;
        cout << "3. push_front(head,val)" << endl;
        cout << "4. push_after(head,key,val)" << endl;
        cin >> ch;
        if (ch == 1)
        {
            cout << "enter val : ";
            int val;
            cin >> val;
            push_back(head, val);
        }
        else if (ch == 2)
        {
            display_list(head);
        }
        else if (ch == 3)
        {
            cout << "enter val : ";
            int val;
            cin >> val;
            push_front(head, val);
        }
        else if (ch == 4)
        {
            cout << "enter val : ";
            int val;
            cin >> val;
            cout << "enter key : ";
            int key;
            cin >> key;
            push_after(head, key, val);
        }
    }
}